src/main/java/pro/trevor/tankgame/state/board/unit/tank/GenericTank.java

```diff
@@ -0,0 +1,158 @@
+package pro.trevor.tankgame.state.board.unit.tank;
+
+import org.json.JSONArray;
+import org.json.JSONObject;
+import pro.trevor.tankgame.rule.type.IPlayerElement;
+import pro.trevor.tankgame.rule.type.ITickElement;
+import pro.trevor.tankgame.state.board.IMovable;
+import pro.trevor.tankgame.state.board.IPositioned;
+import pro.trevor.tankgame.state.board.Position;
+import pro.trevor.tankgame.state.board.unit.IUnit;
+import pro.trevor.tankgame.state.board.unit.tank.status.IAttributeStatus;
+import pro.trevor.tankgame.state.board.unit.tank.status.IStatus;
+import pro.trevor.tankgame.state.board.unit.tank.status.IStatusDecoder;
+
+import java.util.*;
+
+public class GenericTank<E extends Enum<E> & IAttribute> implements IUnit, IMovable, IPositioned, ITickElement, IPlayerElement {
+
+    protected final String player;
+    protected Position position;
+    protected final Map<E, Object> attributes;
+    protected final Set<IStatus> statuses;
+
+    public GenericTank(String player, Position position, Map<E, Object> defaults) {
+        this.player = player;
+        this.position = position;
+        this.attributes = new HashMap<>();
+        this.statuses = new HashSet<>();
+        for (E attribute : defaults.keySet()) {
+            attributes.put(attribute, defaults.get(attribute));
+        }
+    }
+
+    public GenericTank(JSONObject json, IAttributeDecoder<E> attributeDecoder, IStatusDecoder statusDecoder) {
+        assert !json.get("type").equals("tank");
```

Comment:
This assert doesn't need to be here if this was an old tank then the attributes field would be missing.  Plus this means you won't be able to name a class tank in the future.